File: android/live-update-provider/src/test/kotlin/io/ionic/liveupdateprovider/federatedcapacitor/ProviderRegistryTests.kt
package io.ionic.liveupdateprovider.federatedcapacitor

import android.content.Context
import io.ionic.liveupdateprovider.ProviderError
import io.ionic.liveupdateprovider.ProviderManager
import org.junit.Assert.assertEquals
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

class ProviderRegistryTests {
    @Test
    fun `require throws ProviderNotRegistered for unknown id`() {
        val unknownId = uniqueProviderId()

        try {
            ProviderRegistry.require(unknownId)
            throw AssertionError("Expected ProviderNotRegistered")
        } catch (error: ProviderError.ProviderNotRegistered) {
            assertEquals(unknownId, error.providerId)
        }
    }

    @Test
    fun `register and require returns the same provider`() {
        val id = uniqueProviderId()
        val provider = TestProvider(id)

        ProviderRegistry.register(provider)

        assertSame(provider, ProviderRegistry.resolve(id))
        assertSame(provider, ProviderRegistry.require(id))
    }

    @Test
    fun `register throws for empty and blank ids`() {
        val emptyIdProvider = TestProvider("")
        val blankIdProvider = TestProvider("   ")

        try {
            ProviderRegistry.register(emptyIdProvider)
            throw AssertionError("Expected RegistrationFailed for empty ID")
        } catch (error: ProviderError.RegistrationFailed) {
            assertTrue(error.details.contains("empty ID"))
        }

        try {
            ProviderRegistry.register(blankIdProvider)
            throw AssertionError("Expected RegistrationFailed for blank ID")
        } catch (error: ProviderError.RegistrationFailed) {
            assertTrue(error.details.contains("empty ID"))
        }
    }

    @Test
    fun `duplicate registration throws RegistrationFailed`() {
        val id = uniqueProviderId()
        val first = TestProvider(id)
        val second = TestProvider(id)

        ProviderRegistry.register(first)
        try {
            ProviderRegistry.register(second)
            throw AssertionError("Expected RegistrationFailed for duplicate ID")
        } catch (error: ProviderError.RegistrationFailed) {
            assertTrue(error.details.contains("already registered"))
        }
    }

    @Test
    fun `concurrent duplicate registration keeps first and throws for others`() {
        val id = uniqueProviderId()
        val first = TestProvider(id)
        ProviderRegistry.register(first)
        val registrationFailedCount = AtomicInteger(0)

        runConcurrent(times = 24) {
            try {
                ProviderRegistry.register(TestProvider(id))
            } catch (error: ProviderError.RegistrationFailed) {
                registrationFailedCount.incrementAndGet()
            }
        }

        assertSame(first, ProviderRegistry.resolve(id))
        assertEquals(24, registrationFailedCount.get())
    }

    @Test
    fun `concurrent unique registrations are resolvable`() {
        val runId = UUID.randomUUID().toString()
        val providersById = ConcurrentHashMap<String, TestProvider>()

        runConcurrent(times = 60) { index ->
            val id = "provider-$runId-$index"
            val provider = TestProvider(id)
            providersById[id] = provider
            ProviderRegistry.register(provider)
        }

        providersById.forEach { (id, provider) ->
            assertSame(provider, ProviderRegistry.resolve(id))
        }
    }

    private fun uniqueProviderId(): String = "provider-${UUID.randomUUID()}"

    private fun runConcurrent(times: Int, action: (Int) -> Unit) {
        val pool = Executors.newFixedThreadPool(8)
        val done = CountDownLatch(times)

        repeat(times) { index ->
            pool.execute {
                try {
                    action(index)
                } finally {
                    done.countDown()
                }
            }
        }

        val finished = done.await(10, TimeUnit.SECONDS)
        pool.shutdownNow()
        if (!finished) {
            throw AssertionError("Timed out waiting for concurrent operations")
        }
    }

    private class TestProvider(override val id: String) : LiveUpdateProvider {
        override fun createManager(
            context: Context,
            config: Map<String, Any>
        ): ProviderManager {
            throw NotImplementedError("Not needed for registry tests")
        }
    }
}

File: ios/Tests/LiveUpdateProviderTests/ProviderRegistryTests.swift
import XCTest
@testable import LiveUpdateProvider

private struct MockManager: ProviderManager {
    let latestAppDirectory: URL? = nil

    func sync() async throws -> (any ProviderSyncResult)? {
        nil
    }
}

private struct MockProvider: LiveUpdateProvider {
    let id: String

    func createManager(
        config: [String: Any]
    ) throws -> any ProviderManager {
        MockManager()
    }
}

final class ProviderRegistryTests: XCTestCase {
    func testResolveReturnsRegisteredProvider() throws {
        let registry = ProviderRegistry.shared
        let providerId = "test-provider-\(UUID().uuidString)"
        let provider = MockProvider(id: providerId)

        try registry.register(provider)

        let resolvedProvider = registry.resolve(providerId)

        XCTAssertEqual(resolvedProvider?.id, providerId)
    }

    func testResolveReturnsNilForUnknownProvider() {
        let providerId = "missing-provider-\(UUID().uuidString)"

        XCTAssertNil(ProviderRegistry.shared.resolve(providerId))
    }

    func testRequireReturnsRegisteredProvider() throws {
        let registry = ProviderRegistry.shared
        let providerId = "required-provider-\(UUID().uuidString)"

        try registry.register(MockProvider(id: providerId))

        let provider = try registry.require(providerId)

        XCTAssertEqual(provider.id, providerId)
    }

    func testRequireThrowsProviderNotRegisteredForUnknownProvider() {
        let providerId = "missing-provider-\(UUID().uuidString)"

        do {
            _ = try ProviderRegistry.shared.require(providerId)
            XCTFail("Expected require to throw for missing provider")
        } catch ProviderError.providerNotRegistered(let id) {
            XCTAssertEqual(id, providerId)
        } catch {
            XCTFail("Expected providerNotRegistered, got: \(error)")
        }
    }

    func testConcurrentUniqueRegistrationsAreResolvable() async throws {
        let registry = ProviderRegistry.shared
        let runId = UUID().uuidString

        try await withThrowingTaskGroup(of: Void.self) { group in
            for i in 0..<100 {
                group.addTask {
                    let id = "concurrent-provider-\(runId)-\(i)"
                    try registry.register(MockProvider(id: id))
                    _ = registry.resolve(id)
                }
            }

            try await group.waitForAll()
        }

        for i in 0..<100 {
            let id = "concurrent-provider-\(runId)-\(i)"
            XCTAssertEqual(registry.resolve(id)?.id, id)
        }
    }

    func testRegisterThrowsForBlankProviderId() {
        let registry = ProviderRegistry.shared
        let providers = [MockProvider(id: ""), MockProvider(id: "   ")]

        for provider in providers {
            do {
                try registry.register(provider)
                XCTFail("Expected register to throw for blank provider ID")
            } catch ProviderError.registrationFailed(let message) {
                XCTAssertTrue(message.contains("empty ID"))
            } catch {
                XCTFail("Expected registrationFailed, got: \(error)")
            }
        }
    }

    func testRegisterThrowsForDuplicateProviderId() throws {
        let registry = ProviderRegistry.shared
        let providerId = "duplicate-provider-\(UUID().uuidString)"

        try registry.register(MockProvider(id: providerId))

        do {
            try registry.register(MockProvider(id: providerId))
            XCTFail("Expected register to throw for duplicate provider ID")
        } catch ProviderError.registrationFailed(let message) {
            XCTAssertTrue(message.contains("already registered"))
        } catch {
            XCTFail("Expected registrationFailed, got: \(error)")
        }
    }
}
